src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -115,45 +115,47 @@ public static FormulaValue Value(EvalVisitor runner, SymbolContext symbolContext
         // https://docs.microsoft.com/en-us/powerapps/maker/canvas-apps/functions/function-text
         public static FormulaValue Text(EvalVisitor runner, SymbolContext symbolContext, IRContext irContext, FormulaValue[] args)
         {
-            if (args.Length > 1)
+            // only DateValue and DateTimeValue are supported for now with custom format strings.
+            if (args.Length > 1 && args[0] is StringValue)
             {
-                return CommonErrors.NotYetImplementedError(irContext, "Text() doesn't support format args");
+                return CommonErrors.NotYetImplementedError(irContext, "Text() doesn't support format args for type StringValue");
             }
 
-            // $$$ combine with a ToString()? 
-            // $$$ Should these be handled by coercion?
-            string str = null;
-            var arg0 = args[0];
-            if (arg0 is NumberValue num)
-            {
-                str = num.Value.ToString();
-            }
-            else if (arg0 is StringValue s)
-            {
-                str = s.Value;
-            }
-            else if (arg0 is DateValue d)
-            {
-                // $$$ Use real format string
-                str = d.Value.ToString("M/d/yyyy", runner.CultureInfo);
-            }
-            else if (arg0 is DateTimeValue dt)
+            string resultString = null;
+            string formatString = null;
+
+            if (args.Length > 1 && args[1] is StringValue)
             {
-                // $$$ Use real format string
-                str = dt.Value.ToString("g", runner.CultureInfo);
+                formatString = ((StringValue)args[1]).Value;
             }
-            else if (arg0 is TimeValue t)
+
+            switch (args[0])
             {
-                var tDate = new DateTime().Add(t.Value);
-                str = tDate.ToString("t", runner.CultureInfo);
+                case NumberValue num:
+                    resultString = num.Value.ToString(formatString ?? "g", runner.CultureInfo);
+                    break;
+                case StringValue s:
+                    resultString = s.Value;
+                    break;
+                case DateValue d:
+                    resultString = d.Value.ToString(formatString ?? "M/d/yyyy", runner.CultureInfo);
+                    break;
+                case DateTimeValue dt:
+                    resultString = dt.Value.ToString(formatString ?? "g", runner.CultureInfo);
+                    break;
+                case TimeValue t:
+                    resultString = new DateTime().Add(t.Value).ToString(formatString ?? "t", runner.CultureInfo);
```

Comment:
Low priority, but if the format contains a year, this will be different than Excel: `Text(Time(13,35,57), "yyyy-mm hh:mm:ss")` - it will show "1900-01 13:35:57" in Excel, and "0001-01 13:35:57" with this implementation. We should align with the "0-value" for the date (1899-12-31)

Reply:
Right, raw times without a date will be relative to the epoch defined in CoercionUnary.cs

Reply:
So this should really be

`_epoch.Add(t.Value)...`

Reply:
Getting the start value to be in alignment is one issue, and we can fix that pretty easily with a "new DateTime(1900,1,1).Add(t.Value)...  The bigger issues come with the .net runtime is there are differences between "mm" and "MM" and "hh" and "HH". The .net runtime interprets the format string as wanting us to return "mm" or minutes and "hh" as in 12 hour clock hours, so you get back is: "1900-35 01:35:57" and to the .net runtime this is exactly what we asked for based on a format string of "yyyy-mm hh:mm:ss".  We would have to start modifying format strings to replicate the behavior of Excel which is what we were trying to avoid on this simple implementation. Once we get a comprehensive test set that can exercise all these quirks, then we can start to address them knowing we aren't breaking one thing to fix something else.

Reply:
The epoch for our implementation is actually 1899-12-30 as implied mentioned above. This is because Excel incorrectly thinks that the year 1900 is a leap year, but it is not. So our epoch is offset by one day for negative numbers, 0, and the first 60 or so days.